SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/proof/ProofChecker.java

```diff
@@ -3402,22 +3776,74 @@ boolean isApplication(final String funcSym, final Term term) {
 	}
 
 	/**
-	 * Checks if a term is an annotation term with a single annotation without value.
+	 * Checks if a term is an annotation term with a single annotation. Usually the annotation has no value, there are
+	 * some exceptions that are checked.
 	 *
 	 * @param term
 	 *            the term to check.
-	 * @return the annotation key or null if it is not a correct annotation.
+	 * @return the annotation or null if it is not a correct annotation.
 	 */
-	String checkAndGetAnnotationKey(final Term term) {
+	private Annotation getSingleAnnotation(final Term term) {
 		if (term instanceof AnnotatedTerm) {
 			final Annotation[] annots = ((AnnotatedTerm) term).getAnnotations();
-			if (annots.length == 1 && annots[0].getValue() == null) {
-				return annots[0].getKey();
+			if (annots.length == 1) {
+				final Annotation singleAnnot = annots[0];
+				if (singleAnnot.getKey() == ":subst" || singleAnnot.getKey() == ":skolem"
+						|| singleAnnot.getKey() == ":removeForall") {
+					if (singleAnnot.getValue() instanceof Term[]) {
+						return singleAnnot;
+					}
+				} else if (singleAnnot.getValue() == null) {
+					return singleAnnot;
+				}
 			}
 		}
 		return null;
 	}
 
+	/**
+	 * Check that an {@literal @}AUX term has the same definition as previously seen.
+	 */
+	private boolean compareAuxDef(final Term auxTerm, final Term defTerm) {
+		assert auxTerm instanceof ApplicationTerm
+				&& ((ApplicationTerm) auxTerm).getFunction().getName().startsWith("@AUX");
+		final ApplicationTerm auxApp = (ApplicationTerm) auxTerm;
+		for (final Term p : auxApp.getParameters()) {
+			assert p instanceof TermVariable;
+		}
+		if (!mQuantDefinedTerms.containsKey(auxApp)) {
+			mQuantDefinedTerms.put(auxApp, defTerm);
+			return true;
+		} else {
+			return mQuantDefinedTerms.get(auxApp) == defTerm;
+		}
+	}
+
+	/**
+	 * Check that an existentially quantified variable has a unique Skolem function.
+	 * 
+	 * @param sk
+	 * @param existentialVars
+	 * @param subformula
+	 */
+	private boolean compareSkolemDef(Term sk, TermVariable var, Term subformula) {
+		assert sk instanceof ApplicationTerm;
+		final ApplicationTerm skolemApp = (ApplicationTerm) sk;
+		final String func = skolemApp.getFunction().getName();
+		final Term defTerm = mSkript.quantifier(0, new TermVariable[] { var }, subformula);
+		for (final ApplicationTerm a : mQuantDefinedTerms.keySet()) {
```

Comment:
This is inefficient.  Can't we just use mQuantDefinedTerms.get(a), or use the skolem function symbol as key?

Reply:
The idea here was that the function symbol might be used with different arguments (e.g. after DER). But I'm not sure now what happens/should happen in this case.

Reply:
I added a new patch, that also separates the skolem hash map from the aux map.